bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/dto/CloudDTO.java

```diff
@@ -0,0 +1,358 @@
+        return tokenRequestedAt;
+    }
+
+    private HttpClient httpClient;
+
+    /**
+     * Client for Http requests
+     * 
+     * @return httpClient
+     */
+    public HttpClient getHttpClient() {
+        return httpClient;
+    }
+
+    /**
+     * Sets Http Client
+     * 
+     * @param httpClient Http Client
+     */
+    public void setHttpClient(HttpClient httpClient) {
+        this.httpClient = httpClient;
+    }
+
+    private String errMsg;
+
+    /**
+     * Gets error message
+     * 
+     * @return errMsg
+     */
+    public String getErrMsg() {
+        return errMsg;
+    }
+
+    private @Nullable String accessToken = "";
+
+    private String loginAccount;
+    private String password;
+    private CloudProviderDTO cloudProvider;
+    private Security security;
+
+    private @Nullable String loginId;
+    private String sessionId;
+
+    /**
+     * Parameters for Cloud Provider
+     * 
+     * @param email email
+     * @param password password
+     * @param cloudProvider Cloud Provider
+     */
+    public CloudDTO(String email, String password, CloudProviderDTO cloudProvider) {
+        this.loginAccount = email;
+        this.password = password;
+        this.cloudProvider = cloudProvider;
+        this.security = new Security(cloudProvider);
+        logger.debug("Cloud provider: {}", cloudProvider.getName());
+    }
+
+    /**
+     * Set up the initial data payload with the global variable set
+     */
+    private JsonObject apiRequest(String endpoint, JsonObject args, JsonObject data) {
+        if (data == null) {
+            data = new JsonObject();
+            data.addProperty("appId", cloudProvider.getAppId());
+            data.addProperty("format", FORMAT);
+            data.addProperty("clientType", CLIENT_TYPE);
+            data.addProperty("language", LANGUAGE);
+            data.addProperty("src", cloudProvider.getAppId());
+            data.addProperty("stamp", new SimpleDateFormat("yyyyMMddHHmmss").format(new Date()));
+        }
+
+        // Add the method parameters for the endpoint
+        if (args != null) {
+            for (Map.Entry<String, JsonElement> entry : args.entrySet()) {
+                data.add(entry.getKey(), entry.getValue().getAsJsonPrimitive());
+            }
+        }
+
+        // Add the login information to the payload
+        if (!data.has("reqId") && !Objects.isNull(cloudProvider.getProxied())
+                && !cloudProvider.getProxied().isBlank()) {
+            data.addProperty("reqId", Utils.tokenHex(16));
+        }
+
+        String url = cloudProvider.getApiUrl() + endpoint;
+
+        int time = (int) (new Date().getTime() / 1000);
+
+        String random = String.valueOf(time);
+
+        // Add the sign to the header
+        String json = data.toString();
+        logger.debug("Request json: {}", json);
+
+        Request request = httpClient.newRequest(url).method(HttpMethod.POST).timeout(15, TimeUnit.SECONDS);
```

Comment:
This class is named a dto (reference: https://en.wikipedia.org/wiki/Data_transfer_object) but it seems have logic and specific behavior (constructing and performing API-requests). 
I propose to extract the API-request handling to a seperate class and leacve the DTO with the absolute minumum

Reply:
That is going to be project.  I haven't really touched or understand the whole discovery process, it was part of the code that was always working.  TBH the only reason I changed this to DTO was because I could not correct all the null errors when adding @nullByDefault and read in the docs that it wasn't needed on DTO

Reply:
Let's correct the other parts first and we can then look at this again.